flink-cdc-runtime/src/main/java/org/apache/flink/cdc/runtime/operators/transform/PostTransformProcessor.java

```diff
@@ -0,0 +1,184 @@
+    private List<ProjectionColumn> cachedProjectionColumns;
+
+    public PostTransformProcessor(
+            TableInfo tableInfo,
+            TableChangeInfo tableChangeInfo,
+            TransformProjection transformProjection,
+            @Nullable TransformFilter transformFilter,
+            String timezone) {
+        this.tableInfo = tableInfo;
+        this.tableChangeInfo = tableChangeInfo;
+        this.transformProjection = transformProjection;
+        this.transformFilter = transformFilter;
+        this.timezone = timezone;
+        this.projectionColumnProcessorMap = new ConcurrentHashMap<>();
+        this.cachedProjectionColumns = cacheProjectionColumnMap(tableInfo, transformProjection);
+    }
+
+    public boolean hasTableChangeInfo() {
+        return this.tableChangeInfo != null;
+    }
+
+    public boolean hasTableInfo() {
+        return this.tableInfo != null;
+    }
+
+    public static PostTransformProcessor of(
+            TableInfo tableInfo,
+            TransformProjection transformProjection,
+            TransformFilter transformFilter,
+            String timezone) {
+        return new PostTransformProcessor(
+                tableInfo, null, transformProjection, transformFilter, timezone);
+    }
+
+    public static PostTransformProcessor of(
+            TableChangeInfo tableChangeInfo,
+            TransformProjection transformProjection,
+            TransformFilter transformFilter) {
+        return new PostTransformProcessor(
+                null, tableChangeInfo, transformProjection, transformFilter, null);
+    }
+
+    public static PostTransformProcessor of(
+            TransformProjection transformProjection, TransformFilter transformFilter) {
+        return new PostTransformProcessor(null, null, transformProjection, transformFilter, null);
+    }
+
+    public Schema processSchemaChangeEvent(Schema schema) {
+        List<ProjectionColumn> projectionColumns =
+                TransformParser.generateProjectionColumns(
+                        transformProjection.getProjection(), schema.getColumns());
+        transformProjection.setProjectionColumns(projectionColumns);
+        return schema.copy(
+                projectionColumns.stream()
+                        .map(ProjectionColumn::getColumn)
+                        .collect(Collectors.toList()));
+    }
+
+    public BinaryRecordData processData(BinaryRecordData payload, long epochTime) {
+        List<Object> valueList = new ArrayList<>();
+        List<Column> columns = tableInfo.getSchema().getColumns();
+
+        for (int i = 0; i < columns.size(); i++) {
+            if (cachedProjectionColumns.get(i) != null) {
+                ProjectionColumn projectionColumn = cachedProjectionColumns.get(i);
+                projectionColumnProcessorMap.putIfAbsent(
+                        projectionColumn.getColumnName(),
+                        ProjectionColumnProcessor.of(tableInfo, projectionColumn, timezone));
```

Comment:
We still need to calculate hashcode of projectionColumn.getColumnName() here even projectionColumnProcessorMap already contains this key, could we use List to reduce this calculation?

Reply:
Done. Considering `ProjectionColumnProcessor` has carried `ProjectionColumn` inside, I've combined two caches into one single `List`. CMIIW, seems all column processors must be created to correctly handle a `DataChangeEvent`, so the lazy initialization is redundant and has been removed.

Reply:
@melin Sorry but I didn't see any relevance with this PR. Would you mind opening another ticket and provide a minimum reproducible example about your feature request?

Reply:
@melin's issue has been tracked as FLINK-35774.